components/region_cache_memory_engine/src/range_manager.rs

```diff
@@ -0,0 +1,260 @@
+// Copyright 2024 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use engine_traits::CacheRange;
+
+use crate::engine::{RagneCacheSnapshotMeta, SnapshotList};
+
+#[derive(Debug, Default)]
+pub struct RangeMeta {
+    id: u64,
+    range_snapshot_list: SnapshotList,
+    evicted: bool,
+    can_read: bool,
+    safe_ts: u64,
+}
+
+impl RangeMeta {
+    fn new(id: u64) -> Self {
+        Self {
+            id,
+            range_snapshot_list: SnapshotList::default(),
+            evicted: false,
+            can_read: false,
+            safe_ts: 0,
+        }
+    }
+
+    fn derive_from(id: u64, r: &RangeMeta) -> Self {
+        Self {
+            id,
+            range_snapshot_list: SnapshotList::default(),
+            evicted: false,
+            can_read: r.can_read,
+            safe_ts: r.safe_ts,
+        }
+    }
+
+    pub(crate) fn range_snapshot_list(&self) -> &SnapshotList {
+        &self.range_snapshot_list
+    }
+}
+
+#[derive(Default)]
+struct IdAllocator(u64);
```

Comment:
Either document that this should be called under a lock, or use an atomic.

Reply:
The whole RangeManager is protected by a lock.